File: Python/Bit Manipulation/E-338. Counting Bits.py
class Solution:
    @classmethod
    def countBits(self, n: int) -> list[int]:
        result = [0] * (n + 1)
        for i in range(1, n + 1):
            result[i] = result[i >> 1] + (1 & i)
        return result


testData = [
    {"n": 2, "ans": [0, 1, 1]},
    # {"n": 5, "ans": [0, 1, 1, 2, 1, 2]},
]

for data in testData:
    print(data["ans"] == Solution.countBits(data["n"]))

File: Bit Manipulation/E-338. Counting Bits/E-338. Counting Bits.ts
function countBits(n: number): number[] {
  const result = new Array(n + 1).fill(0);
  for (let i = 1; i <= n; i++) {
    result[i] = result[i >> 1] + (i & 1);
  }
  return result;
}

const testData = [
  { n: 2, ans: [0, 1, 1] },
  { n: 5, ans: [0, 1, 1, 2, 1, 2] },
];

for (const data of testData) {
  console.log(JSON.stringify(data.ans) === JSON.stringify(countBits(data.n)));
}
